powertools-parameters/src/main/java/software/amazon/lambda/powertools/parameters/BaseProvider.java

```diff
@@ -0,0 +1,195 @@
+     * If for some parameters, you need to set a different maxAge, use {@link #withMaxAge(int)}.
+     *
+     * @param maxAgeinSec Maximum time (in seconds) to cache the parameter, before calling the underlying parameter store.
+     * @return the provider itself in order to chain calls (eg. <code>provider.defaultMaxAge(10).get("key")</code>).
+     */
+    public <T extends BaseProvider> T defaultMaxAge(int maxAgeinSec) {
+        this.defaultMaxAge = maxAgeinSec;
+        this.maxAge = maxAgeinSec;
+        return (T) this;
+    }
+
+    /**
+     * (Optional) Builder method to call before {@link #get(String)} or {@link #get(String, Class)}
+     * to set cache max age (in seconds) for the parameter to get.<br/><br/>
+     * The max age is reset to default (either 5 or a custom value set with {@link #defaultMaxAge}) after each get,
+     * so you need to use this method for each parameter to cache with non-default max age.<br/><br/>
+     *
+     * <b>Not Thread Safe</b>: calling this method simultaneously by several threads
+     * can lead to unwanted cache time for some parameters.<br/>
+     *
+     * @param maxAgeinSec Maximum time (in seconds) to cache the parameter, before calling the underlying parameter store.
+     * @return the provider itself in order to chain calls (eg. <code>provider.withMaxAge(10).get("key")</code>).
+     */
+    public <T extends BaseProvider> T withMaxAge(int maxAgeinSec) {
+        this.maxAge = maxAgeinSec;
+        return (T) this;
+    }
+
+    /**
+     * Builder method to call before {@link #get(String)} (Optional) or {@link #get(String, Class)} (Mandatory).
+     * to provide a {@link Transformer} that will transform the String parameter into something else (String, Object, ...)<br/><br/>
+     *
+     * {@link software.amazon.lambda.powertools.parameters.transform.Base64Transformer} and {@link software.amazon.lambda.powertools.parameters.transform.JsonTransformer}
+     * are provided for respectively base64 and json content. You can also write your own (see {@link Transformer}).
+     *
+     * <b>Not Thread Safe</b>: calling this method simultaneously by several threads
+     * can lead to errors (one Transformer for the wrong target type)<br/>
+     *
+     * @param transformerClass Class of the transformer to apply. For convenience, you can use {@link Transformer#json} or {@link Transformer#base64} shortcuts.
+     * @return the provider itself in order to chain calls (eg. <code>provider.withTransformation(json).get("key", MyObject.class)</code>).
+     */
+    public <T extends BaseProvider> T withTransformation(Class<? extends Transformer> transformerClass) {
+        this.transformerClass = transformerClass;
+        return (T) this;
+    }
+
+    /**
+     * Abstract method to tell to retrieve parameters recursively. Only available for {@link SSMProvider}.
+     * @param <T>
+     * @return
+     */
+    public <T extends BaseProvider> T recursive() {
```

Comment:
recursive is only available for SSM when getting multiple parameters. I don't want to force to implement a useless method in other child class (and do what: throw a not implemented exception ?)

Reply:
Yes, I would have preferred that instead. In most of the cases, users will be using the providers that are built in the library like SSM etc and thats probably one of the reason they would want to use this utility to avoid creating one themselves.

On the other hand if this is SSM specific method, do we even need it on BaseProvider then ?

Reply:
Not sure to get your preference. For sure, throwing a not implemented exception is a really bad practice (against SOLID principles). Eventually put a log to notify the developer the method is useless in that case.
I've added it to the Base in order to be able to chain all calls. It was only in SSM at the beginning, but then you cannot do provider.recursive().withMaxAge().get() or provider.withMaxAge().recursive().get()

Reply:
aha i see the reason why its there in base class then .  

Hmm may be its ok the way its is rite now then. I dont see any real value in adding log either.

Reply:
I am wondering at this point, what advantage do we get by introducing inheritance here with  BaseProvider ? 

What would the implementation looks like if these were concrete classes on its own ? How much code is duplicated vs complexity it introduces ? 

For clients perspective, if I understand correctly , they still end up calling `ParamManager` static method for getting instance of param provider of their choice, what advantage inheritance gives us here 🤔

Reply:
The BaseProvider handles the cache mechanism + transformation. It was done just like that in the Python version and it looks good to me. I don't want to duplicate that in several implementations. Moreover if we add DynamoDB after and potentially others. But you are right that it breaks the composition over inheritance pattern... let me think how we could avoid those empty methods and minimize code duplication...

Reply:
Yeah. As long as code duplication is concerned, we can always extract the entire caching logic into a method which can be resued by any provider.